Stackdriver: Deep linking from Grafana panels to the Metrics Explorer

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
A new link `View in Metrics Exploer` is added in the graph panel which opens the Metrics explorer in the Monitoring and runs the query from the Grafana panel there.
![cloud-moniitoring-deep-linking](https://user-images.githubusercontent.com/1632407/85840871-eb191d80-b7a5-11ea-9282-06e15d8e42af.png)


Setup:
- Credentials can be found in the [private plugin-provisioning repo](https://github.com/grafana/plugin-provisioning/blob/master/plugins/stackdriver/provisioning/datasources/stackdriver.yml)
- Provision the YAML by placing it in the conf/provisioning/datasources and restarting your Grafana instance
- [Sample](https://gist.github.com/papagian/69a5fe2c1864204251e1c8c5684966af) dashboard

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #25669

**Special notes for your reviewer**:
The link is only available for metrics queries (not for SLOs).
The user should be logged with a google account (otherwise is navigated to the login page first) that has access to the GCP project specified in the query (otherwise gets a permissions error).

A complementary PR adding a new image for the documentation will follow up.

papagian committed:
Add stackdriver deep link

papagian committed:
Fix linting issue

papagian committed:
No deep link for SLO queries

papagian committed:
Add tests

papagian committed:
Some improvements

papagian committed:
Fix link to the doc image

papagian committed:
Fix misspelling

marefr commented on docs/sources/features/datasources/stackdriver.md:
````diff
@@ -323,3 +323,12 @@ datasources:
       authenticationType: gce
 ```
+
+## Deep linking from Grafana panels to the Metrics Explorer in Gooogle Monitoring
````
Not sure if cloud should be included? 
```suggestion
## Deep linking from Grafana panels to the Metrics Explorer in Gooogle Cloud Monitoring
```

papagian replied:
I'm not completely sure about this; in the google console is referred only as Monitoring.

marefr replied:
That's fine. Probably me not understanding that you're referring to the console with "Gooogle Monitoring".

marefr commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -323,3 +323,12 @@ datasources:
+
+{{< docs-imagebox img="/img/docs/v71/cloud-moniitoring-deep-linking.png" max-width="500px" class="docs-image--right" caption="Google Cloud Monitoring deep linking" >}}
+
+Click on a time series in the panel to see a context menu with a link to View in Monitoring Metrics explorer. Clicking that link opens the Metrics explorer in the Monitoring and runs the query from the Grafana panel there.
```
Not sure about this either.
```suggestion
Click on a time series in the panel to see a context menu with a link to View in Cloud Monitoring Metrics explorer. Clicking that link opens the Metrics explorer in the Cloud Monitoring and runs the query from the Grafana panel there.
```

marefr commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -323,3 +323,12 @@ datasources:
+{{< docs-imagebox img="/img/docs/v71/cloud-moniitoring-deep-linking.png" max-width="500px" class="docs-image--right" caption="Google Cloud Monitoring deep linking" >}}
+
+Click on a time series in the panel to see a context menu with a link to View in Monitoring Metrics explorer. Clicking that link opens the Metrics explorer in the Monitoring and runs the query from the Grafana panel there.
+If you're not currently logged in to the Google console, then the link opens the login page. The provided link is valid for any account, but it only displays the query if your account has access to the GCP project specified in the query.
```
Maybe a note about deep linking for SLO queries not supported?

marefr reviewed (changes requested):
Great. Haven't tested it a lot, but expect you to do more of that. Added a couple comment/suggestions.  Please open website PR with image update so we can test docs before merging this .

Was a bit hard just clicking on link when you have multiple google accounts and it selected the wrong one when clicking on link. Had to copy it and append `&authuser=1` :smile:

papagian commented:
> Was a bit hard just clicking on link when you have multiple google accounts and it selected the wrong one when clicking on link. Had to copy it and append `&authuser=1` 😄

so do you suggest to include this in the url?
if we do, the user somehow should configure this; I don't think that selecting always the primary account would work for everyone.

papagian committed:
Update stackdriver.md

papagian commented:
> Please open website PR with image update so we can test docs before merging this .

https://github.com/grafana/website/pull/1724

oddlittlebird commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -323,3 +323,14 @@ datasources:
+
+{{< docs-imagebox img="/img/docs/v71/cloud-moniitoring-deep-linking.png" max-width="500px" class="docs-image--right" caption="Google Cloud Monitoring deep linking" >}}
+
+> Available only for Metric queries
```
```suggestion
> **Note:** This feature is only available only for Metric queries.
```

papagian committed:
Update docs/sources/features/datasources/stackdriver.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

daniellee reviewed (approved):
In the Metrics Explorer, when you create a new query, the first value you choose is the resource type. A list of resource types is not available in the public API and hence, we had to build it a different way in Grafana (Service + metric is the first choice). This means you often get this error message about a resource not have being selected:

![image](https://user-images.githubusercontent.com/434655/85981954-4f75f000-b9e5-11ea-9889-8caa59163631.png)

If the user adds resource type to their filter then this error disappears:
![image](https://user-images.githubusercontent.com/434655/85982113-9532b880-b9e5-11ea-9c26-5f308945c9a4.png)

Not sure if we can do anything about this. There could theoretically be multiple resource types so adding it automatically to the query is probably not a good idea. Maybe can mention it in the docs.

Otherwise, everything seems to work as expected. (Though I also have the problem with multiple accounts which I also have with Google Meet, Google Docs etc.)

papagian committed:
Update stackdreiver docs

papagian committed:
Enforce resource type

papagian committed:
Navigate to google account chooser first

papagian commented:
> > Was a bit hard just clicking on link when you have multiple google accounts and it selected the wrong one when clicking on link. Had to copy it and append `&authuser=1` 😄
> 
> so do you suggest to include this in the url?
> if we do, the user somehow should configure this; I don't think that selecting always the primary account would work for everyone.

Following [this](https://github.com/grafana/grafana/pull/25874#issuecomment-651168276) comment's suggestion I found [this post](https://github.com/grafana/grafana/pull/25874#issuecomment-651168276) and I ended up with the solution to direct first to the AccountChooser (the `prompt` parameter - suggested in the post discussion - didn't work in this case because I suspect that it is only supported in OAuth2 authorization URLs).

papagian commented:
> In the Metrics Explorer, when you create a new query, the first value you choose is the resource type. A list of resource types is not available in the public API and hence, we had to build it a different way in Grafana (Service + metric is the first choice). This means you often get this error message about a resource not have being selected:
> 
> Not sure if we can do anything about this. There could theoretically be multiple resource types so adding it automatically to the query is probably not a good idea. Maybe can mention it in the docs.
> 

The resource type information for a given metric type is included in the Monitoring API response. Given a metric type having multiple resource types is considered a rare case we can enforce the first resource type found in the response to be included in the filter parameter (if no resource type is already provided in the query filters).

papagian committed:
Update stackdriver.md

papagian committed:
Change renamed image reference

Fix misspelling in image name and change it to conform with the rest cloud monitoring images

marefr reviewed (approved):
LGTM

Docs look good. It works good. 

It's a little bit weird that I have to go through the account chooser every time, but I can live with that since it's a much better experience than before. However, I wonder what the behavior is if you only have 1 google account - will it redirect directly to target link? Tried to find documentation about google account chooser, but came up empty, where did you find documentation?

papagian commented:
> It's a little bit weird that I have to go through the account chooser every time, but I can live with that since it's a much better experience than before. However, I wonder what the behavior is if you only have 1 google account - will it redirect directly to target link?

No, you still pass first from the account chooser. Another option would be to add a new configuration and only if it's enabled to enforce account chooser.

> Tried to find documentation about google account chooser, but came up empty, where did you find documentation?

There is a reference [here](https://developers.google.com/identity/toolkit/web/account-chooser#the_account_chooser) but I read about it in the stackoverflow post mentioned above.

papagian committed:
Merge remote-tracking branch 'origin/master' into 25669-stackdriver-deep-linking

papagian committed:
Merge remote-tracking branch 'origin/master' into 25669-stackdriver-deep-linking

papagian committed:
Update cloudmonitoring.md

daniellee reviewed (approved):
I tested with one account and it was smooth. Just one extra click that went really fast. Adding the resource type works well too.

LGTM

papagian committed:
Merge remote-tracking branch 'origin/master' into 25669-stackdriver-deep-linking

oddlittlebird commented on docs/sources/features/datasources/cloudmonitoring.md:
````diff
@@ -324,3 +324,14 @@ datasources:
       authenticationType: gce
 ```
+
+## Deep linking from Grafana panels to the Metrics Explorer in Gooogle Cloud Console
````
```suggestion
## Deep linking from Grafana panels to the Metrics Explorer in Google Cloud Console
```

oddlittlebird commented on docs/sources/features/datasources/cloudmonitoring.md:
```diff
@@ -324,3 +324,14 @@ datasources:
+
+> **Note:** This feature is only available for Metric queries.
+
+Click on a time series in the panel to see a context menu with a link to View in Metrics explorer in Google Cloud Console. Clicking that link opens the Metrics explorer in the Google Cloud Console and runs the query from the Grafana panel there.
+The link navigates the user first to the Google Account Chooser and after successfully selecting an account, the user is redirected to the metrics explorer. The provided link is valid for any account, but it only displays the query if your account has access to the GCP project specified in the query.
```
Inconsistent capitalization of "metrics explorer" here. If it is a proper noun (named in the UI), then it should be capitalized to match the UI. If it is not, then lowercase throughout. Either way, try to be consistent.

papagian replied:
```suggestion
Click on a time series in the panel to see a context menu with a link to View in Metrics Explorer in Google Cloud Console. Clicking that link opens the Metrics Explorer in the Google Cloud Console and runs the query from the Grafana panel there.
The link navigates the user first to the Google Account Chooser and after successfully selecting an account, the user is redirected to the Metrics Explorer. The provided link is valid for any account, but it only displays the query if your account has access to the GCP project specified in the query.
```

oddlittlebird reviewed (commented):
Got a couple suggestions, looks good otherwise.

papagian committed:
Apply suggestions from code review

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

ericvolp12 commented:
Hey, I've noticed that with a Stat panel type, there is no way to disable this form of deep linking and any click on the panel takes you to the deep link. We'd like to be able to link a stat to another dashboard in our Grafana for drill-downs but can't do that because every stat panel forcibly links to GCP and even with a Data Link configured for the panel, it gets overridden by the GCP deep link. I would love to just disable GCP deep links for a panel or even a whole dashboard if possible.
